Upgrade to Hibernate Search 6.0.0 beta3

yrodiere opened the pull request:
https://in.relation.to/2019/12/16/hibernate-search-6-0-0-Beta3/

@gsmet I will need your advice, especially on the reflection stuff. I'm not sure I fully grasped what was going on there, so I may have removed important code.

To summarize, starting with Beta3 all Hibernate Search mapping annotations are meta-annotated, so we do not need to list them all. Custom annotations must also be meta-annotated with the same meta-annotation, so they are no longer a specific use case.

However, it seems we were behaving differently for built-in annotations and custom bridge annotations (the code wasn't exactly the same). I did as well as I could, but that may not be enough...

yrodiere committed:
Upgrade to Hibernate Search 6.0.0.Beta3

yrodiere committed:
Filter out the new "version" log from Hibernate Search on startup

gsmet commented:
CI is unhappy: the native build is failing with some weird logging issue:
```
Exception in thread "main" java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:231)
	at io.quarkus.runtime.Application.start(Application.java:87)
	at io.quarkus.runtime.Application.run(Application.java:210)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:41)
Caused by: javax.persistence.PersistenceException: [PersistenceUnit: default] Unable to build Hibernate SessionFactory
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.persistenceException(FastBootEntityManagerFactoryBuilder.java:113)
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:67)
	at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:54)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
	at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:109)
	at io.quarkus.hibernate.orm.runtime.JPAConfig.startAll(JPAConfig.java:57)
	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.startAllPersistenceUnits(HibernateOrmRecorder.java:82)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits59.deploy_0(HibernateOrmProcessor$startPersistenceUnits59.zig:51)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits59.deploy(HibernateOrmProcessor$startPersistenceUnits59.zig:70)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:161)
	... 3 more
Caused by: java.lang.IllegalArgumentException: Invalid logger interface org.hibernate.search.engine.logging.impl.Log (implementation not found in java.lang.ClassLoader@3477f70)
	at org.jboss.logging.Logger$1.run(Logger.java:2556)
	at java.security.AccessController.doPrivileged(AccessController.java:80)
	at org.jboss.logging.Logger.getMessageLogger(Logger.java:2529)
	at org.jboss.logging.Logger.getMessageLogger(Logger.java:2516)
	at org.hibernate.search.util.common.logging.impl.LoggerFactory.make(LoggerFactory.java:21)
	at org.hibernate.search.engine.Version.logVersion(Version.java:32)
	at org.hibernate.search.mapper.orm.bootstrap.impl.HibernateSearchIntegrator.integrate(HibernateSearchIntegrator.java:65)
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:286)
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:462)
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:65)
	... 12 more
```
Did you change something regarding logging?

yrodiere commented:
@gsmet This part of the stack is new in Beta3:

```
	at org.hibernate.search.engine.Version.logVersion(Version.java:32)
```

However, I do not understand why it's failing. I would have expected reflection to be enabled for all JBoss loggers as part of the jboss-logger Quarkus extension.

Do I have to make sure all loggers are instantiated during deployment? If so, we may have a problem, because most loggers are instantiated when a class is first loaded, and I'm pretty sure we have several other classes using a logger that are not used until runtime (`SimpleIndexingProgressMonitor` being the most obvious example I have in mind).

gsmet commented:
@dmlloyd I think we need your advice here.

yrodiere committed:
Add more assertions for mass indexing in Hibernate Search

dmlloyd commented:
We do not enable reflection in any logging classes that I am aware of.  This kind of problem happens when loggers are not created in static initializers, which should always be the case.

We did a fairly significant cleanup effort around this a while ago for performance reasons.  I think if Hibernate Search has started using non-constant loggers again, I would consider it a regression in that project for that reason alone.  Also: does Hibernate Search *really* need its own logging abstraction?

That said, we do need a good substitution strategy for loggers which support i18n, because I believe right now it will pin the locale of the building process rather than the run time environment.

yrodiere commented:
> We did a fairly significant cleanup effort around this a while ago for performance reasons. I think if Hibernate Search has started using non-constant loggers again, I would consider it a regression in that project for that reason alone.

Looking closer, there is indeed a behavior change: the logger used to be created (then dropped, to be garbage collected) in a static intiailizer, and is now initialized at runtime (then dropped).

Not sure it's a performance regression, though, since the logger is only instantiated once (just as before).

> Also: does Hibernate Search really need its own logging abstraction?

As a rule, no. `Version.logVersion` is an exception. I guess I can get rid of it.

yrodiere commented:
Created https://hibernate.atlassian.net/browse/HSEARCH-3795 to address this. I'll see what I can do in this PR to make it work with `6.0.0.Beta3`.

Thanks for the insight, @dmlloyd .

yrodiere committed:
Work around a problem with Version.logVersion in Hibernate Search 6.0.0.Beta3

gsmet commented on extensions/hibernate-search-elasticsearch/deployment/src/main/java/io/quarkus/hibernate/search/elasticsearch/HibernateSearchLogFilterBuildStep.java:
```diff
@@ -0,0 +1,16 @@
+
+    @BuildStep
+    void setupLogFilters(BuildProducer<LogCleanupFilterBuildItem> filters) {
+        filters.produce(new LogCleanupFilterBuildItem("org.hibernate.search.engine.Version", "HSEARCH000034"));
```
I wouldn't make it a separate class. We tend to centralize these things in the processor.

yrodiere replied:
Ok, I just copy-pasted from someplace else. Will do.

gsmet commented on extensions/hibernate-search-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/elasticsearch/runtime/HibernateSearchElasticsearchRecorder.java:
```diff
@@ -124,6 +124,8 @@ private void contributeBackendRuntimeProperties(BiConsumer<String, Object> prope
             addBackendConfig(propertyCollector, backendName, ElasticsearchBackendSettings.HOSTS,
                     elasticsearchBackendConfig.hosts);
+            addBackendConfig(propertyCollector, backendName, ElasticsearchBackendSettings.PROTOCOL,
+                    elasticsearchBackendConfig.protocol.getHibernateSearchString());
```
Can you add a note here: https://github.com/quarkusio/quarkus/wiki/Migration-Guide ?

yrodiere replied:
As a matter of fact, I cannot. I don't have the sufficient rights.

Here's something, if you want to add it yourself:

````
### Hibernate Search + Elasticsearch (Preview)

The protocol used to contact Elasticsearch is now specified in a separate configuration property, `quarkus.hibernate-search.elasticsearch.protocol`. This property defaults to `http`. 

As a result, the property `quarkus.hibernate-search.elasticsearch.discovery.default_scheme` was removed.

So if your configuration looked like this:

```
quarkus.hibernate-search.elasticsearch.hosts = http://es1.mycompany.com
```

You should now have this:

```
quarkus.hibernate-search.elasticsearch.hosts = es1.mycompany.com
```

And if it looked like this:

```
quarkus.hibernate-search.elasticsearch.hosts = https://es1.mycompany.com
quarkus.hibernate-search.elasticsearch.discovery.enabled = true
quarkus.hibernate-search.elasticsearch.discovery.default_scheme = https
```

You should now have this:

```
quarkus.hibernate-search.elasticsearch.hosts = es1.mycompany.com
quarkus.hibernate-search.elasticsearch.protocol = https
quarkus.hibernate-search.elasticsearch.discovery.enabled = true
```
````

gsmet replied:
Added, thanks.

gsmet reviewed (changes requested):
Added two minor comments.

The reflection changes looks good. I was maybe a bit too conservative before but with meta annotations, you don't have a choice.

yrodiere commented:
@gsmet I addressed your comments.
I also added some assertions in the mass indexing test following our discussion (where I was worried that maybe, mass indexing wasn't effective -- spoiler alert: it works).

gsmet commented:
Merged, thanks!
